### DISCUSSION TITLE
Regular refresh of token required

### DISCUSSION THREAD

**Discussion by iandarbey** [tsightler/ring-mqtt#411] [Support Q&A] [answered]
Hi,

Following a recent update I've had to refresh my login token on a regular basis. Is this intended behaviour or should I remove and reinstall?

It's a bit frustrating as I've automations run off the door sensors which randomly don't work if the token is erroring out.

Thanks

**Comment by tsightler** [comment]
This is not expected behavior, I have systems that have been running for over a year without having to manually refresh the token.  I just fired up my development instance for the first time in a month and it logged right in with the saved token.  The token is automatically refreshed every hour and the updated token is saved in the state file for use during new logins.  If you are using the Docker install method you must have a persistent volume mapped to save this state.

Note that there are things that might make tokens become invalid, like changing the account password, or removing the device from authorized devices (there's generally two devices one "unknown" and another "ring-mqtt" or "ring-mqtt-addon").  Either of these actions will expire the token immediately.

**Reply by whinson** [reply]
I am having the same problem.  

I went from 4.6.3 -> Current Version and started running into this issue.  With 4.6.3 it worked without problem.

Now after a random interval of time (sometimes 30 minutes, sometimes half a day - but, generally shorter than longer), I run into an issue where the saved token seemingly doesn't work.  In my experience, bouncing the container resolves the issue 90% of the time.  When it doesn't, stopping the container completely and going through the initial setup step to generate the token resolves the issue. 

The one caveat to all that is my HD died on me right before a trip, so I was down for a couple weeks.  After going through the restore process to get all my containers back-up, I am no longer seeing the error message related to the token in the logs (kind of confused on that) that I was seeing before the HD crash and I don't have the error message saved.  The behavior is the same though, where all my sensors go offline and the resolution steps work the same - so I assume it is ultimately the same error I was seeing previously. 

I suspect the issue is stemming somehow from the ring-client-api library, but I don't see any issues opened or relevant discussions over there and there is nothing I can pinpoint to say for certain.  And I haven't been able to figure out the right log settings to get a better picture of what may be happening.

I am running in docker and I am happy to provide any other needed information/enable any needed logging to get to the bottom of this.  Or if you have any other suggestions, I am happy to try those.

Thanks in Advance.  I really appreciate this project and the value it adds to my home automation setup.

**Reply by tsightler** [reply]
There have been no significant changes between 4.6.3 -> Current versions regarding token management, other than the fact that tokens are no longer allowed in the configuration (this was already the case for addon and for Docker when using a persistent mount).  Since you say restarting the container solves it 90% of the time then it seems unlikely it's actually a token issue for those cases at least, but logs are the only way to be sure.

I doubt the issue is in ring-client-api, but more likely some unique combination of environmental factors that triggers the behavior.  Between this project and other projects built on ring-client-api, there are 1000's of users of ring-client-api (my estimate for this project alone is ~1500 users, based on stars, github and docker hub pull rates, etc, and ring-homebridge is a much more popular project).

Also, I run 3 different instances of this project myself, and know personally about a dozen people who use it and would not be shy about reporting such issues to me, and, last time I asked, none of them have generated a token in months.  I have literally not even restarted my instances since I upgraded to 5.0.2 and it continues to work.  Any time there is any common issue with tokens expiring regularly I get dozens of complaints a week, both here, on HA forums, and on Reddit posts however, this discussion post is basically the first mention of it in many months, so I don't think it is a common problem plaguing many users. 

That being said, it's of course possible, heck even likely, that bugs still exist and I'm certainly not saying there is no possible problems with tokens, but there's zero reason for tokens to only last 30 minutes or even one day, as tokens are long-lived.  I know that ring-client-api had some weird issues with tokens for users running it on MacOS and I'm not sure they were ever resolved.  If someone wants to open an issue and provide complete details of their configuration as well as full debug logs, I'm certainly willing to look into it.  If it really happens commonly in half a day, or less, it should be quite easy to find the root cause.

**Reply by whinson** [reply]
I am happy to provide the information/do whatever is needed.  

Just to make sure I provide everything needed for the logs, you want me to set logging to "DEBUG=*".  When I did that previously, I got a bunch of RTSP/MQTT logging that I am not sure is useful for this problem.  I am also concerned with the amount of logging being produced.  Ultimately - happy to whatever is needed.

And then for the complete details of the configuration, you want my docker-compose? What other information would be helpful?

One other random thought that I am not sure is helpful or not at this point:
The issue will sometimes resolve itself without any intervention from myself.  In the below picture from the last day, you can see it was successfully able to reconnect a couple times without any action from me:

<img width="196" alt="image" src="https://user-images.githubusercontent.com/11203642/182492591-1a628ffa-6f3e-4e2c-820b-ed5e3452aa3b.png">

**Reply by whinson** [reply]
Well I turned on full logging and I lost the connection in a just a few minutes:  

Below is what looks to be the issue, though I have no idea why the server would randomly time out.  Other thing to point out is that scrolling up through the logs, it looks like this is maybe the 5th retry or so and on this one, it sent the 'Offline' MQTT message.  

Quickly scrolling, nothing else stood out to me as a cause for the timeout in the log and checking my pi-hole logs, I see those URLs at least made it there and were NOT blocked at that point.  I can upload the full file log when I open the issue as requested.

Thanks!


```
2022-08-03T00:14:41.083Z ring [31mFailed to reach Ring server at https://app-snaps.ring.com/snapshots/next/87944002.  Timeout awaiting 'request' for 20000ms.  Trying again in 5 seconds...[39m
2022-08-03T00:14:41.086Z ring [31mFailed to reach Ring server at https://api.ring.com/clients_api/doorbots/87944002/health.  Timeout awaiting 'request' for 20000ms.  Trying again in 5 seconds...[39m
2022-08-03T00:14:41.094Z ring [31mFailed to reach Ring server at https://app-snaps.ring.com/snapshots/next/128063545.  Timeout awaiting 'request' for 20000ms.  Trying again in 5 seconds...[39m
2022-08-03T00:14:41.095Z ring [31mFailed to reach Ring server at https://api.ring.com/clients_api/doorbots/128063545/health.  Timeout awaiting 'request' for 20000ms.  Trying again in 5 seconds...[39m
2022-08-03T00:14:41.096Z ring [31mFailed to reach Ring server at https://app-snaps.ring.com/snapshots/next/88185120.  Timeout awaiting 'request' for 20000ms.  Trying again in 5 seconds...[39m
2022-08-03T00:14:41.098Z ring [31mFailed to reach Ring server at https://api.ring.com/clients_api/doorbots/88185120/health.  Timeout awaiting 'request' for 20000ms.  Trying again in 5 seconds...[39m
2022-08-03T00:14:41.102Z ring [31mFailed to reach Ring server at https://app-snaps.ring.com/snapshots/next/88639142.  Timeout awaiting 'request' for 20000ms.  Trying again in 5 seconds...[39m
2022-08-03T00:14:41.105Z ring [31mFailed to reach Ring server at https://api.ring.com/clients_api/doorbots/88639142/health.  Timeout awaiting 'request' for 20000ms.  Trying again in 5 seconds...[39m
2022-08-03T00:14:44.608Z ring-mqtt Websocket for location id bb5..... is disconnected
2022-08-03T00:14:44.608Z ring-mqtt [32m[ Bridge] [39m[34mring/bb5.../status [39m[36moffline[39m
2022-08-03T00:14:44.609Z mqttjs:client publish :: message `offline` to topic `ring/bb5.../alarm/4...4e7/status`
2022-08-03T00:14:44.609Z mqttjs:client publish :: qos 1
2022-08-03T00:14:44.609Z mqttjs:client MqttClient:publish: packet cmd: publish
2022-08-03T00:14:44.609Z mqttjs:client _sendPacket :: (mqttjs_b706d22c) ::  start
2022-08-03T00:14:44.609Z mqttjs:client storeAndSend :: store packet with cmd publish to outgoingStore
2022-08-03T00:14:44.610Z mqttjs:client nop :: undefined
2022-08-03T00:14:44.610Z mqttjs:client sendPacket :: packet: {
```

**Reply by tsightler** [reply]
```
2022-08-03T00:14:41.102Z ring �[31mFailed to reach Ring server at https://app-snaps.ring.com/snapshots/next/88639142.  Timeout awaiting 'request' for 20000ms.  Trying again in 5 seconds...�[39m
2022-08-03T00:14:41.105Z ring �[31mFailed to reach Ring server at https://api.ring.com/clients_api/doorbots/88639142/health.  Timeout awaiting 'request' for 20000ms.  Trying again in 5 seconds...�[39m
2022-08-03T00:14:44.608Z ring-mqtt Websocket for location id bb5..... is disconnected
```

This is the critical part and doesn't appear to have anything to do with the token, if it did, it would show the request being rejected due to being unauthorized.  Instead, it's clearly showing that the inability to reach the Ring server API endpoints via the network.  Note that it is failing to receive responses from three completely different API endpoints, app-snaps.ring.com, api.ring.com, and also the websocket connection was closed as well.  This is telling because these methods use two completely different Ring APIs and connect using different methods, the first two are API endpoints for cameras which use standard https requests via the got client library, while the websocket uses socket.io for the connection.  The camera APIs and socket APIs are completely unrelated and use totally separate code paths, yet both are failing indicating network connectivity issues.

Unless there is something earlier in the logs that show it crashed somehow, I have to conclude that this is a network issue as these errors are far below ring-mqtt/ring-client-api.

**Reply by whinson** [reply]
Thank you for looking!

In an effort to not chase network issues, I have done the following today:

- I stood up a second instance of Ring-MQTT at version 4.7.3 (as that is the last label I see) on the same box/everything else the sa,e.  
- I hacked the code for the 4.7.3 version simply to append 'OLD' onto the deviceID in the base constructor to give me a different topic and ultimately different entities in HA to get easier/better visibility into if the connection persists as opposed to monitoring logs/MQTT.

So, I effectively have two set of Ring MQTT instances and associated HA entities - one at the 5.0.2 revision and one at the 4.7.3 - both pointing to my same Ring Account.  

With that, the 4.7.3 version's connection has consistently persisted (no issues thus far) whereas the 5.0.2 connection has acted as detailed above.  This matches with what I would expect based on my memory of upgrading - as far as I remember, the only thing that changed before the issue started happening was the Ring-MQTT image and needed configuration changes. 

So, my mind jumps to wondering if my issues stems from the APIs changing between versions and ultimately my system not handling the new APIs as well.  It just doesn't make sense to me that there would be a network issue with the same set of APIs between different images hosted on the same machine with the same network configurations.  

All that being said, not knowing the ins and outs of the code very well, I am sure there is some other more plausible root cause I am missing.

At this point, I see 3 paths forward (assuming 4.7.3 being stable for me continues to be a true statement):

1. Slowly work forward on versions to see when/if it breaks and hone in on the changes in the library dependencies/etc. 
2. Figure out better logging to understand what is happening with the connection to Ring - there seems to be minimal logging devoted to the actual ring connection in the 'enable all' logging (e.g. All I see is timeouts, none of the successful connections to understand what the APIs are when it works) and to your point above, that seems like an issue below your code/the ring-client-api.
3. Move on with our lives as 4.7.3 is good enough for me and again as you've stated, I seem to be a weird edge case.

I am more than happy to help with options 1 and 2 or any other approach to figuring out the issue, but I realize I can only take it so far and would need a good bit of support, which you may not want to provide/have time to provide.  

So with that, I am perfectly content to stay at 4.7.3 and we can all move on with our lives.

Please let me know if it is worth it to try any of the above suggestions or anything else, otherwise I'll plan on moving on with my life.

Thanks!

**Reply by tsightler** [reply]
Ideally I think it would be best to start a new thread since this is clearly not related to token management (at least, based on the logs you've provided so far).  I mean, failed to reach Ring server and disconnecting websocket are pretty low level, even below ring-client-api.  I've been thinking about this problem and the only thing that comes to mind is perhaps some issue with the Docker image itself.  I've seen some network related weirdness reported in Alpine images later than 3.12, and current Docker image is using Alpine 3.15.4.  What Docker versions are you running and on what OS?  Do you have IPv6 addresses on your equipment?

The reality is, you are welcome to continue to use 4.7.3 until it stops working.  That could be tomorrow, or it might be years from now, who knows, but Ring is pretty much known to make changes to the API which break things on ring-client-api at least once a year, so, at some point, 4.7.3 won't be good enough, but for now if that's what you'd like to go with, it's certainly a fine option.

**Comment by iandarbey** [comment, marked as answer]
Yes this has gone a bit off topic.

For clarity - I deleted the add-on from home assistant and reinstalled and setup from scratch.

Issue appears resolved - would suggest anyone having similar issues that it's some form of issue with an upgrade of the add on or home assistant (I'm running on a RPi 4) and that a delete and reinstall of the add on should resolve.

**Comment by tlemmons** [comment]
I have been having this problem.  I have to regenerate often.  Sometimes a few days, sometimes a few minutes.  I have tried to find the json config file on my system (virtualbox vm) and cannot find it anywhere on the vm file system.  Not sure where to look.  I have reinstalled the plugin a couple of times and nothing seems to stop the eventual failure.   I am willing to do anything I can to help.
